WalletWasabi/Services/UpdateManager.cs

```diff
@@ -0,0 +1,276 @@
+using Newtonsoft.Json.Linq;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+using WalletWasabi.Microservices;
+using WalletWasabi.Models;
+using WalletWasabi.Tor.Http;
+
+namespace WalletWasabi.Services;
+
+public class UpdateManager : IDisposable
+{
+	private string InstallerPath { get; set; } = "";
+	private const byte MaxTries = 3;
+	private const string ReleaseURL = "https://api.github.com/repos/zkSNACKs/WalletWasabi/releases/latest";
+
+	public UpdateManager(string dataDir, bool downloadNewVersion, IHttpClient httpClient)
+	{
+		InstallerDir = Path.Combine(dataDir, "Installer");
+		HttpClient = httpClient;
+		DownloadNewVersion = downloadNewVersion;
+	}
+
+	private async void UpdateChecker_UpdateStatusChangedAsync(object? sender, UpdateStatus updateStatus)
+	{
+		var tries = 0;
+		bool updateAvailable = !updateStatus.ClientUpToDate || !updateStatus.BackendCompatible;
+		Version targetVersion = updateStatus.ClientVersion;
+		if (!updateAvailable)
+		{
+			Cleanup();
+			return;
+		}
+		if (DownloadNewVersion)
+		{
+			do
+			{
+				tries++;
+				try
+				{
+					(string installerPath, Version newVersion) = await GetInstallerAsync(targetVersion).ConfigureAwait(false);
+					InstallerPath = installerPath;
+					Logger.LogInfo($"Version {newVersion} downloaded successfuly.");
+					updateStatus.IsReadyToInstall = true;
+					updateStatus.ClientVersion = newVersion;
+					break;
+				}
+				catch (Exception ex)
+				{
+					Logger.LogError($"Geting new update failed with error.", ex);
+				}
+			} while (tries < MaxTries);
+		}
+
+		UpdateAvailableToGet?.Invoke(this, updateStatus);
+	}
+
+	/// <summary>
+	/// Get or download installer for the newest release.
+	/// </summary>
+	/// <param name="targetVersion">This does not contains the revision number, because backend always sends zero.</param>
+	private async Task<(string filePath, Version newVersion)> GetInstallerAsync(Version targetVersion)
+	{
+		(Version newVersion, string url, string fileName) = await GetLatestReleaseFromGithubAsync(targetVersion).ConfigureAwait(false);
+
+		var tmpFilePath = Path.Combine(InstallerDir, $"{fileName}.tmp");
+		var newFilePath = Path.Combine(InstallerDir, fileName);
+
+		var installerDownloaded = TryGetDownloadedInstaller(fileName);
+		if (!installerDownloaded)
+		{
+			EnsureToRemoveCorruptedFiles();
+
+			// This should also be done using Tor.
+			// TODO: https://github.com/zkSNACKs/WalletWasabi/issues/8800
+			Logger.LogInfo($"Trying to download new version: {newVersion}");
+			using HttpClient httpClient = new();
```

Comment:
This is a clearnet HttpClient only used here. How else should this be added?
Create a ClearnetHttpClient with HttpClientFactory and send it down in the constructor?

Reply:
Yeah I hear you. Right now I think it's good this way, if it will be changed in the future anyway to work through tor. Later replacing it like as it is now will be easy.

Reply:
It should be static then, according to the pattern.